src/beast/beast/crypto/impl/ripemd_context.h

```diff
@@ -0,0 +1,441 @@
+
+    // parallel round 5
+    BEAST_RIPEMD_FFF(bbb, ccc, ddd, eee, aaa, X[12] ,  8);
+    BEAST_RIPEMD_FFF(aaa, bbb, ccc, ddd, eee, X[15] ,  5);
+    BEAST_RIPEMD_FFF(eee, aaa, bbb, ccc, ddd, X[10] , 12);
+    BEAST_RIPEMD_FFF(ddd, eee, aaa, bbb, ccc, X[ 4] ,  9);
+    BEAST_RIPEMD_FFF(ccc, ddd, eee, aaa, bbb, X[ 1] , 12);
+    BEAST_RIPEMD_FFF(bbb, ccc, ddd, eee, aaa, X[ 5] ,  5);
+    BEAST_RIPEMD_FFF(aaa, bbb, ccc, ddd, eee, X[ 8] , 14);
+    BEAST_RIPEMD_FFF(eee, aaa, bbb, ccc, ddd, X[ 7] ,  6);
+    BEAST_RIPEMD_FFF(ddd, eee, aaa, bbb, ccc, X[ 6] ,  8);
+    BEAST_RIPEMD_FFF(ccc, ddd, eee, aaa, bbb, X[ 2] , 13);
+    BEAST_RIPEMD_FFF(bbb, ccc, ddd, eee, aaa, X[13] ,  6);
+    BEAST_RIPEMD_FFF(aaa, bbb, ccc, ddd, eee, X[14] ,  5);
+    BEAST_RIPEMD_FFF(eee, aaa, bbb, ccc, ddd, X[ 0] , 15);
+    BEAST_RIPEMD_FFF(ddd, eee, aaa, bbb, ccc, X[ 3] , 13);
+    BEAST_RIPEMD_FFF(ccc, ddd, eee, aaa, bbb, X[ 9] , 11);
+    BEAST_RIPEMD_FFF(bbb, ccc, ddd, eee, aaa, X[11] , 11);
+
+    // combine results
+    ddd += cc + ctx.h[1]; // final result for h[0]
+    ctx.h[1] = ctx.h[2] + dd + eee;
+    ctx.h[2] = ctx.h[3] + ee + aaa;
+    ctx.h[3] = ctx.h[4] + aa + bbb;
+    ctx.h[4] = ctx.h[0] + bb + ccc;
+    ctx.h[0] = ddd;
+}
+
+template <class = void>
+void init (ripemd160_context& ctx) noexcept
+{
+    ctx.len = 0;
+    ctx.tot_len = 0;
+    ctx.h[0] = 0x67452301UL;
+    ctx.h[1] = 0xefcdab89UL;
+    ctx.h[2] = 0x98badcfeUL;
+    ctx.h[3] = 0x10325476UL;
+    ctx.h[4] = 0xc3d2e1f0UL;
+}
+
+template <class = void>
+void update (ripemd160_context& ctx,
+    void const* message, std::size_t size) noexcept
+{
+    auto const pm = reinterpret_cast<
+        unsigned char const*>(message);
+    unsigned int block_nb;
+    unsigned int new_len, rem_len, tmp_len;
+    const unsigned char *shifted_message;
+    tmp_len = ripemd160_context::block_size - ctx.len;
+    rem_len = size < tmp_len ? size : tmp_len;
+    std::memcpy(&ctx.block[ctx.len], pm, rem_len);
+    if (ctx.len + size < ripemd160_context::block_size) {
+        ctx.len += size;
+        return;
+    }
+    new_len = size - rem_len;
+    block_nb = new_len / ripemd160_context::block_size;
+    shifted_message = pm + rem_len;
+    std::array<std::uint32_t, 16> X;
+    ripemd_load(X, ctx.block);
+    ripemd_compress(ctx, X);
+    for (int i = 0; i < block_nb; ++i)
+    {
+        ripemd_load(X, shifted_message +
+            i * ripemd160_context::block_size);
+        ripemd_compress(ctx, X);
+    }
+    rem_len = new_len % ripemd160_context::block_size;
+    std::memcpy(ctx.block, &shifted_message[
+        block_nb * ripemd160_context::block_size],
+            rem_len);
+    ctx.len = rem_len;
+    ctx.tot_len += (block_nb + 1) *
+        ripemd160_context::block_size;
+}
+
+template <class = void>
+void finish (ripemd160_context& ctx,
+    void* digest) noexcept
+{
+    std::array<std::uint32_t, 16> X;
+    X.fill(0);
+    // put leftovers into X
+    unsigned char const* p =
```

Comment:
1. prefer`auto`
2. Why not declare `ctx.block` to have its actual type instead of having to `reinterpret_cast` it every time?